### PR TITLE
AKS: add node pool configuration

### PR DESCRIPTION
**Pull request by to266**
Adds ability to configure AKS node pools.

~~(at this point not actually tested, therefore marked as draft. Nevertheless, review still very welcome. Will update the status (and rebase+push if needed) once I confirmed that all is good).~~

### CODE REVIEW

File: azurerm/cluster/node_pool/variables.tf
Comment (to266) [inline azurerm/cluster/node_pool/variables.tf:21]: Somehow I was not able to get this to work with the below configuration. Any advice is welcome, but as it is right now results in correct plan.

```
  main = {
    apps = {
      node_pool_name      = "main"
      enable_auto_scaling = true
      max_count           = 9
      min_count           = 3
      node_count          = 3
      vm_size             = "Standard_D2_v3"
      availability_zones  = "1,2,3"
      node_labels         = "component=main"
      os_disk_type        = "Ephemeral"
    }

    ops = {
      min_count          = 1
      node_count         = 1
      max_count          = 5
      availability_zones = ""
      use_spot = true
    }
  }

```

File: azurerm/cluster/node_pool/versions.tf
Comment (to266) [inline azurerm/cluster/node_pool/versions.tf:9]: If no strict type, this is of course not required and would remove

File: azurerm/_modules/aks/node_pool/main.tf
Comment (to266) [inline azurerm/_modules/aks/node_pool/main.tf:25]: I remember a while back having issues with `node_count` when autoscaling, as it's basically used for creation, but also used to be aligned when updating.

Ideally, azure provider should know that when autoscaling is on -> no need to pay attention to the actual node count.

As I never had fixed node pool, I used to have the `node_count` here, but right now I'm unsure if we want this.

File: azurerm/cluster/node_pool/variables.tf
Comment (pst) [reply azurerm/cluster/node_pool/variables.tf:21]: The reason why I use the object for the GKE and EKS node pool modules is, that without it all values have to be strings. That makes it hard to implement e.g. labels and taints.

The thing to be aware of about the optional object attributes, and I assume this is what caused the issue for you, is that they will by default equal null. So if you use lookup on one of the attributes, it won't ever use the default value because the attribute exists but its value is null.

For attributes where you want to pass the null value through, you can therefor simply do `local.cfg["attribute_name"]` but for maps/lists that you may be using in a `for_each` or similar, you will have to default them to empty map or set. In those cases, I found it best to manually check for null with the ternary operator and then handcraft my default value. E.g. `taints = local.cfg["taints"] != null ? local.cfg["taints"] : {}` (assuming taints is a map)

You can see examples of this [here - EKS](https://github.com/kbst/terraform-kubestack/blob/master/aws/cluster/node-pool/configuration.tf#L32) or [here - GKE](https://github.com/kbst/terraform-kubestack/blob/master/google/cluster/node-pool/configuration.tf#L34).

**Review by pst — COMMENTED** [review]
I know this is an inconsistency with the module under `azurerm/_modules/aks/` but  I used dashes instead of underscores for the public facing node pool modules, e.g. `google/cluster/node-pool` and `aws/cluster/node-pool`. This is a consequence of the `elb-dns` module using a dash due to an oversight of mine.

I would prefer to have consistency for all public facing modules (the `*/cluster/*` ones), so it would be great to have the aks module use dash instead of underscore as well.

File: azurerm/_modules/aks/node_pool/main.tf
Comment (pst) [reply azurerm/_modules/aks/node_pool/main.tf:25]: For EKS something similar was required, I also think it should be handled in the provider then but it isn't the case. The docs there say to use lifecycle ignore_changes. I will leave this note here for me to test this later.

File: azurerm/cluster/node_pool/variables.tf
Comment (to266) [reply azurerm/cluster/node_pool/variables.tf]: Thanks for the excellent pointer, now all resolved.

**Comment by to266** [comment]
Renamed as requested. From my tests, works as expected, so all done on my end.

**Comment by pst** [comment]
Thanks, @to266, I'll review and merge this tomorrow, assuming I find no blockers in my testing.

File: azurerm/_modules/aks/node_pool/variables.tf
Comment (pst) [inline azurerm/_modules/aks/node_pool/variables.tf:8]:
```suggestion
  description = "Resource group of the cluster."
```

File: azurerm/_modules/aks/node_pool/variables.tf
Comment (pst) [inline azurerm/_modules/aks/node_pool/variables.tf:48]:
```suggestion
  description = "VM size to use for the nodes in the pool"
```

File: azurerm/_modules/aks/node_pool/variables.tf
Comment (pst) [inline azurerm/_modules/aks/node_pool/variables.tf:19]:
```suggestion
  description = "Whether to enable auto scaling for the node pool. Defaults to true"
```

File: azurerm/_modules/aks/node_pool/variables.tf
Comment (pst) [inline azurerm/_modules/aks/node_pool/variables.tf:71]:
```suggestion
  description = "Whether to use spot instances. Defaults false."
```

File: azurerm/_modules/aks/node_pool/variables.tf
Comment (pst) [inline azurerm/_modules/aks/node_pool/variables.tf:77]:
```suggestion
  description = "The maximum desired spot price, or -1 for the current on-demand price. Defaults to '-1'"
```

File: azurerm/cluster/node-pool/variables.tf
Comment (pst) [inline azurerm/cluster/node-pool/variables.tf:39]:
```suggestion
  description = "The resource group of the cluster to attach the node pool to"
```

File: azurerm/cluster/node-pool/variables.tf
Comment (pst) [inline azurerm/cluster/node-pool/variables.tf:34]:
```suggestion
  description = "The name of the cluster to attach the node pool to"
```

File: azurerm/_modules/aks/node_pool/main.tf
Comment (pst) [inline azurerm/_modules/aks/node_pool/main.tf:1]: After applying the node pool configuration for the first time, a second apply wants to replace the node pool. `vnet_subnet_id` change forces a replacement.

```
# module.aks_zero_node_pool.module.node_pool.azurerm_kubernetes_cluster_node_pool.current must be replaced
-/+ resource "azurerm_kubernetes_cluster_node_pool" "current" {
      - availability_zones     = [] -> null
      - enable_host_encryption = false -> null
      - enable_node_public_ip  = false -> null
      - fips_enabled           = false -> null
      ~ id                     = "/subscriptions/803f877a-b0ff-49a7-afc7-eea135679b9a/resourcegroups/terraform-kubestack-testing/providers/Microsoft.ContainerService/managedClusters/kbstacctest-ops-westeurope/agentPools/test1" -> (known after apply)
      ~ kubelet_disk_type      = "OS" -> (known after apply)
        name                   = "test1"
      ~ node_labels            = {} -> (known after apply)
      ~ orchestrator_version   = "1.20.9" -> (known after apply)
      ~ os_disk_size_gb        = 128 -> (known after apply)
      ~ os_sku                 = "Ubuntu" -> (known after apply)
      - tags                   = {} -> null
      - vnet_subnet_id         = "/subscriptions/803f877a-b0ff-49a7-afc7-eea135679b9a/resourceGroups/terraform-kubestack-testing/providers/Microsoft.Network/virtualNetworks/kbstacctest-ops-westeurope/subnets/kbstacctest-ops-westeurope-default-node-pool" -> null # forces replacement
        # (14 unchanged attributes hidden)
    }
```

It's probably best, to make this attribute configurable, but default to the value of the [cluster's data source](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/data-sources/kubernetes_cluster#vnet_subnet_id).

So something like this in the configuration.tf locals:

```hcl
vnet_subnet_id = local.cfg["vnet_subnet_id"] != null ? local.cfg["vnet_subnet_id"] : data.azurerm_kubernetes_cluster.current.vnet_subnet_id
```

File: azurerm/_modules/aks/node_pool/main.tf
Comment (pst) [inline azurerm/_modules/aks/node_pool/main.tf:10]: Currently, you set defaults in different ways in multiple locations. Over time, I came to the conclusion, that it's best to only implement the defaults in the configuration.tf locals. And set no defaults for the inner module's variables. This ensures, that the outer module calls the inner module with all available variables.

In your case, the third layer of logic to set defaults is the conditionals here around the variables. It may be worth giving this a brief though, if it can be refactored to all be done in configuration.tf just once.

I don't consider this an issue that would block merging. Just a small improvement for future maintainability. But I have some tech-debt related to this in all the modules. So if you don't have time to clean this up right now, I am ok with adding it to the list for later.

**Review by pst — CHANGES_REQUESTED** [review]
This looks really good overall. The default value suggestion is not critical. I left a few smaller suggestions for typos I found.

The only topic blocking the merge is the `vnet_subnet_id` issue. Once that's addressed, I am happy to merge. Thanks for putting in the work to contribute this.

File: azurerm/_modules/aks/node_pool/main.tf
Comment (pst) [reply azurerm/_modules/aks/node_pool/main.tf:1]: One thought, I don't know if it makes sense to have different node pools use different subnet IDs on Azure. It's a thing on AWS. So assuming it does not make sense on Azure, you might as well just set it to what the cluster data source returns.

Comment (to266) [reply azurerm/_modules/aks/node_pool/main.tf:1]: At least on the [terraform side](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/kubernetes_cluster_node_pool#vnet_subnet_id) they have this:
 > At this time the vnet_subnet_id must be the same for all node pools in the cluster

So I'll not expose the config, but rather just take it from the cluster data source

File: azurerm/_modules/aks/node_pool/main.tf
Comment (to266) [reply azurerm/_modules/aks/node_pool/main.tf]: Tried to do it, as seemed straightforward enough. Let me know if that's not what you really meant :)

File: azurerm/_modules/aks/node_pool/main.tf
Comment (to266) [reply azurerm/_modules/aks/node_pool/main.tf:1]: Turned out a bit less straightforward, as (at least in my configuration) all the `vnet_subnet_id` values that were returned by the data source were actually empty strings. Seems to be to do the sensible thing, though.

```
❯ terraform state show data.azurerm_kubernetes_cluster.current
# data.azurerm_kubernetes_cluster.current:
data "azurerm_kubernetes_cluster" "current" {
    addon_profile                   = [
 # ...snip
    ]
    agent_pool_profile              = [
        {
            availability_zones       = [
                "1",
                "2",
                "3",
            ]
            count                    = 1
            enable_auto_scaling      = true
            enable_node_public_ip    = false
            max_count                = 5
            max_pods                 = 110
            min_count                = 1
            name                     = "main"
            node_labels              = {
                "component" = "main"
            }
            node_public_ip_prefix_id = ""
            node_taints              = []
            orchestrator_version     = "1.21.2"
            os_disk_size_gb          = 128
            os_type                  = "Linux"
            tags                     = {}
            type                     = "VirtualMachineScaleSets"
            upgrade_settings         = []
            vm_size                  = "Standard_D8ds_v4"
            vnet_subnet_id           = ""
        },
        {
            availability_zones       = []
            count                    = 1
            enable_auto_scaling      = true
            enable_node_public_ip    = false
            max_count                = 2
            max_pods                 = 110
            min_count                = 1
            name                     = "system"
            node_labels              = {}
            node_public_ip_prefix_id = ""
            node_taints              = [
                "CriticalAddonsOnly=true:NoSchedule",
            ]
            orchestrator_version     = "1.21.2"
            os_disk_size_gb          = 30
            os_type                  = "Linux"
            tags                     = {}
            type                     = "VirtualMachineScaleSets"
            upgrade_settings         = []
            vm_size                  = "Standard_B4ms"
            vnet_subnet_id           = ""
        },
        {
            availability_zones       = [
                "1",
                "2",
                "3",
            ]
            count                    = 0
            enable_auto_scaling      = true
            enable_node_public_ip    = false
            max_count                = 3
            max_pods                 = 110
            min_count                = 0
            name                     = "worker"
            node_labels              = {
                "component" = "bi-worker"
            }
            node_public_ip_prefix_id = ""
            node_taints              = []
            orchestrator_version     = "1.21.2"
            os_disk_size_gb          = 128
            os_type                  = "Linux"
            tags                     = {}
            type                     = "VirtualMachineScaleSets"
            upgrade_settings         = []
            vm_size                  = "Standard_D8ds_v4"
            vnet_subnet_id           = ""
        },
    ]
    api_server_authorized_ip_ranges = []
 # ...snip
```

**Review by pst — APPROVED** [review]

**Comment by pst** [comment]
:1st_place_medal: @to266

**Pull request merged**
